tpm/tpm.go

```diff
@@ -628,6 +631,125 @@ func MakeIdentity(rw io.ReadWriter, srkAuth []byte, ownerAuth []byte, aikAuth []
+	if err := binary.Read(symbuf, binary.BigEndian, &header); err != nil {
+		return nil, err
+	}
+	parms := make([]byte, header.Parmsize)
+	if err := binary.Read(symbuf, binary.BigEndian, parms); err != nil {
+		return nil, err
+	}
+	cred := make([]byte, header.Credsize)
+	if err := binary.Read(symbuf, binary.BigEndian, cred); err != nil {
+		return nil, err
+	}
+	return cred, nil
+}
+
+// ActivateIdentity sks the TPM to decrypt an EKPub encrypted symmetric session key
+// which it uses to decrpyt the symmetrically encrypted secret.
+func ActivateIdentity(rw io.ReadWriter, aikAuth []byte, ownerAuth []byte, aik tpmutil.Handle, asym, sym []byte) ([]byte, error) {
+	// Run OIAP for the AIK
+	oiaprAIK, err := oiap(rw)
+	if err != nil {
+		return nil, fmt.Errorf("failed to start OIAP session: %v", err)
+	}
+
+	// Run OSAP for the Owner, reading a random OddOSAP for our initial command
+	// and getting back a secret and a handle.
+	sharedSecretOwn, osaprOwn, err := newOSAPSession(rw, etOwner, khOwner, ownerAuth)
+	if err != nil {
+		return nil, fmt.Errorf("failed to start OSAP session: %v", err)
+	}
+	defer osaprOwn.Close(rw)
+	defer zeroBytes(sharedSecretOwn[:])
+
+	authIn := []interface{}{ordActivateIdentity, asym}
+	ca1, err := newCommandAuth(oiaprAIK.AuthHandle, oiaprAIK.NonceEven, aikAuth, authIn)
+	if err != nil {
+		return nil, fmt.Errorf("newCommandAuth failed: %v", err)
+	}
+	ca2, err := newCommandAuth(osaprOwn.AuthHandle, osaprOwn.NonceEven, sharedSecretOwn[:], authIn)
+	if err != nil {
+		return nil, fmt.Errorf("newCommandAuth failed: %v", err)
+	}
+
+	symkey, ra1, ra2, ret, err := activateIdentity(rw, aik, asym, ca1, ca2)
+	if err != nil {
+		return nil, fmt.Errorf("activateIdentity failed: %v", err)
+	}
+
+	// Check response authentication.
+	raIn := []interface{}{ret, ordActivateIdentity, symkey}
+	if err := ra1.verify(ca1.NonceOdd, aikAuth, raIn); err != nil {
+		return nil, fmt.Errorf("aik resAuth failed to verify: %v", err)
+	}
+
+	if err := ra2.verify(ca2.NonceOdd, sharedSecretOwn[:], raIn); err != nil {
+		return nil, fmt.Errorf("owner resAuth failed to verify: %v", err)
+	}
+
+	cred, err := unloadTrspiCred(sym)
+	var (
+		block     cipher.Block
+		iv        []byte
+		ciphertxt []byte
+		secret    []byte
+	)
+	switch id := symkey.AlgID; id {
+	case algAES128:
+		block, err = aes.NewCipher(symkey.Key)
+		if err != nil {
+			return nil, fmt.Errorf("aes.NewCipher failed: %v", err)
+		}
+		iv = cred[:aes.BlockSize]
+		ciphertxt = cred[aes.BlockSize:]
+		secret = ciphertxt
+	default:
+		return nil, fmt.Errorf("%v is not a supported session key algorithm", id)
+	}
+	switch es := symkey.EncScheme; es {
+	case esSymCTR:
+		stream := cipher.NewCTR(block, iv)
+		stream.XORKeyStream(secret, ciphertxt)
+	case esSymOFB:
+		stream := cipher.NewOFB(block, iv)
+		stream.XORKeyStream(secret, ciphertxt)
+	case esSymCBCPKCS5:
+		mode := cipher.NewCBCDecrypter(block, iv)
+		mode.CryptBlocks(secret, ciphertxt)
+		// remove PKCS5 padding
```

Comment:
nit: please make comments into full sentences, with an initial capital and final punctuation.